src/lib.rs

```diff
@@ -17,244 +17,147 @@
 )]
 #![forbid(unsafe_code)]
 
-#[cfg(feature = "std")]
-#[macro_use]
-extern crate std;
+use ark_ff::{FpParameters, PrimeField};
+use ark_std::vec::Vec;
+use ark_std::vec;
 
-#[cfg(not(feature = "std"))]
 #[macro_use]
-extern crate alloc as std;
+extern crate derivative;
+
+/// Infrastructure for the constraints counterparts.
+#[cfg(feature = "r1cs")]
+pub mod constraints;
+
+mod absorbable;
+pub use absorbable::*;
 
-use ark_ff::models::{
-    Fp256, Fp256Parameters, Fp320, Fp320Parameters, Fp384, Fp384Parameters, Fp768, Fp768Parameters,
-    Fp832, Fp832Parameters,
-};
-use ark_ff::{to_bytes, PrimeField, ToConstraintField};
-use std::vec::Vec;
+/// The sponge for Poseidon
+///
+/// This implementation of Poseidon is entirely from Fractal's implementation in [COS20][cos]
+/// with small syntax changes.
+///
+/// [cos]: https://eprint.iacr.org/2019/1076
+pub mod poseidon;
+
+/// A sponge that offers backwards compatibility for implementations that do not accept sponge
+/// objects but require domain separation. Operates in the same way as fork.
+pub mod domain_separated;
 
 /// An enum for specifying the output field element size.
-#[derive(Clone)]
+#[derive(Clone, Copy, Eq, PartialEq)]
 pub enum FieldElementSize {
     /// Sample field elements from the entire field.
     Full,
 
-    /// Sample field elements from a subset of the field.
-    Truncated {
-        /// The maximum size of the subset is 2^num_bits.
-        num_bits: usize,
-    },
+    /// Sample field elements from a subset of the field, specified by the maximum number of bits.
+    Truncated(usize),
+}
+
+impl FieldElementSize {
+    pub(crate) fn num_bits<F: PrimeField>(&self) -> usize {
+        if let FieldElementSize::Truncated(num_bits) = self {
+            *num_bits.min(&(F::Params::CAPACITY as usize))
```

Comment:
Would it make more sense to return an error if the specified num_bits is greater than the capacity of the field?

Reply:
Yeah that makes sense, because if the protocol using the sponge expects challenges of a particular size, getting smaller challenges might lead to loss of security

 (Though I'd compare with `F::MODULUS_BITS`, not `F::CAPACITY`).

Reply:
Done. btw, why `MODULUS_BITS` instead of `CAPACITY`? @Pratyush